django/core/management/commands/shell.py

```diff
@@ -111,10 +115,52 @@ def python(self, options):
         # Start the interactive interpreter.
         code.interact(local=imported_objects)
 
+    def get_and_report_namespace(self, verbosity):
+        namespace = self.get_namespace()
+
+        if verbosity == 1:
+            self.stdout.write(
+                f"{len(namespace)} objects imported automatically",
+                self.style.SUCCESS,
+            )
+        elif verbosity >= 2:
+            imports_by_module = {}
+            for obj_name, obj in namespace.items():
+                if (
+                    hasattr(obj, "__module__")
+                    and hasattr(obj, "__qualname__")
+                    and obj.__qualname__.find(".") == -1
+                ):
+                    module = obj.__module__
+                    collected_imports = imports_by_module.get(module, [])
+                    imports_by_module[module] = collected_imports + [obj_name]
+            for module, imported_objects in imports_by_module.items():
+                self.stdout.write(
+                    f"from {module} import {', '.join(imported_objects)}",
+                    self.style.SUCCESS,
+                )
+
+        return namespace
+
+    def get_namespace(self):
+        apps_models = apps.get_models()
+        apps_models_modules = [
+            (app.models_module, app.label)
+            for app in apps.get_app_configs()
+            if app.models_module is not None
+        ]
+        namespace = {}
+        for model in reversed(apps_models):
+            if model.__module__:
+                namespace[model.__name__] = model
+        for app_models_module, app_label in apps_models_modules:
+            namespace[f"{app_label}_models"] = app_models_module
```

Comment:
Have you considered the edge case where this key might already exist if someone has (very questionably) named a model `MyApp_models`? I don't think we'd want to overwrite with MyApp's models module in that case.

Reply:
Good Point! This is an easy fix I think